lib/srv/app/gcp/handler.go

```diff
@@ -0,0 +1,276 @@
+	client, err := s.CloudClients.GetGCPIAMClient(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	request := &credentialspb.GenerateAccessTokenRequest{
+		// expected format: projects/-/serviceAccounts/{ACCOUNT_EMAIL_OR_UNIQUEID}
+		Name:  fmt.Sprintf("projects/-/serviceAccounts/%v", serviceAccount),
+		Scope: scopes,
+	}
+	accessToken, err := client.GenerateAccessToken(ctx, request)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return accessToken, nil
+}
+
+// Forwarder is an GCP CLI proxy service that forwards the requests to GCP API, but updates the authorization headers
+// based on user identity.
+type handler struct {
+	// config is the handler configuration.
+	HandlerConfig
+
+	// fwd is used to forward requests to GCP API after the handler has rewritten them.
+	fwd *forward.Forwarder
+
+	// tokenCache caches access tokens.
+	tokenCache *utils.FnCache
+}
+
+// NewGCPHandler creates a new instance of a http.Handler for GCP requests.
+func NewGCPHandler(ctx context.Context, config HandlerConfig) (http.Handler, error) {
+	return newGCPHandler(ctx, config)
+}
+
+// newGCPHandler creates a new instance of a handler for GCP requests. Used by NewGCPHandler and in tests.
+func newGCPHandler(ctx context.Context, config HandlerConfig) (*handler, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	tokenCache, err := utils.NewFnCache(utils.FnCacheConfig{
+		TTL:     time.Second * 60,
+		Clock:   config.Clock,
+		Context: ctx,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	svc := &handler{
+		HandlerConfig: config,
+		tokenCache:    tokenCache,
+	}
+
+	fwd, err := forward.New(
+		forward.RoundTripper(config.RoundTripper),
+		forward.ErrorHandler(oxyutils.ErrorHandlerFunc(svc.formatForwardResponseError)),
+		// Explicitly passing false here to be clear that we always want the host
+		// header to be the same as the outbound request's URL host.
+		forward.PassHostHeader(false),
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	svc.fwd = fwd
+	return svc, nil
+}
+
+// RoundTrip handles incoming requests and forwards them to the proper API.
+func (s *handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	if err := s.serveHTTP(w, req); err != nil {
+		s.formatForwardResponseError(w, req, err)
+		return
+	}
+}
+
+// serveHTTP is a helper to simplify error handling in ServeHTTP.
+func (s *handler) serveHTTP(w http.ResponseWriter, req *http.Request) error {
+	sessionCtx, err := common.GetSessionContext(req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	fwdRequest, err := s.prepareForwardRequest(req, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	recorder := httplib.NewResponseStatusRecorder(w)
+	s.fwd.ServeHTTP(recorder, fwdRequest)
+	status := uint32(recorder.Status())
```

Comment:
It looks like in each case of `ecorder.Status` use the return value if casted to uint32.  Can we change the type returned by recorder.Status ?

Reply:
You could argue that `ResponseStatusRecorder.Status()` should return int because this is consistent with constants in `http` package. Then again, this type was refactored in https://github.com/gravitational/teleport/pull/19387 by @GavinFrazar, and the `uint32` vs. `int` was discussed therein. Any thoughts on this @GavinFrazar ?

Reply:
I think we could change the type for recorder status, it really shouldn't matter. The http package in go uses `int` for status code, as far as I can tell, for simplicity. They don't ever return a negative integer, but they also don't need the extra range of positive values provided by `uint32`. I will open a PR refactoring this. You can just leave it as-is for now.

Reply:
Ok, thanks!